test/unit-test/FreeRTOS_Sockets/FreeRTOS_Sockets_utest.c

```diff
@@ -151,3 +218,7366 @@ void test_FreeRTOS_inet_pton4( void )
     TEST_ASSERT_EQUAL( pdFAIL, xResult );
     TEST_ASSERT_EQUAL_MESSAGE( ulInValidResponse, ulIPAddress, "Incorrectly converted string 8." );
 }
+
+void test_prvValidSocket_InvalidOrNULLSocket( void )
+{
+    BaseType_t xReturn;
+    FreeRTOS_Socket_t * pxSocket = ( FreeRTOS_Socket_t * ) FREERTOS_INVALID_SOCKET;
+    BaseType_t xProtocol, xIsBound;
+
+    xReturn = prvValidSocket( NULL, xProtocol, xIsBound );
+
+    TEST_ASSERT_EQUAL( pdFALSE, xReturn );
+
+    xReturn = prvValidSocket( pxSocket, xProtocol, xIsBound );
+
+    TEST_ASSERT_EQUAL( pdFALSE, xReturn );
+}
+
+void test_prvValidSocket_SocketBoundSetButNotBound( void )
+{
+    BaseType_t xReturn;
+    FreeRTOS_Socket_t xSocket;
+    BaseType_t xProtocol, xIsBound = pdTRUE;
+
+    memset( &xSocket, 0, sizeof( FreeRTOS_Socket_t ) );
+
+    listLIST_ITEM_CONTAINER_ExpectAndReturn( &( xSocket.xBoundSocketListItem ), NULL );
+
+    xReturn = prvValidSocket( &xSocket, xProtocol, xIsBound );
+
+    TEST_ASSERT_EQUAL( pdFALSE, xReturn );
+}
+
+void test_prvValidSocket_SocketBoundResetButBound( void )
+{
+    BaseType_t xReturn;
+    FreeRTOS_Socket_t xSocket;
+    BaseType_t xProtocol, xIsBound = pdFALSE;
+
+    memset( &xSocket, 0, sizeof( FreeRTOS_Socket_t ) );
+
+    xSocket.ucProtocol = xProtocol;
+
+    xReturn = prvValidSocket( &xSocket, xProtocol, xIsBound );
+
+    TEST_ASSERT_EQUAL( pdTRUE, xReturn );
+}
+
+void test_prvValidSocket_InvalidProtocol( void )
+{
+    BaseType_t xReturn;
+    FreeRTOS_Socket_t xSocket;
+    BaseType_t xProtocol, xIsBound = pdTRUE;
+
+    memset( &xSocket, 0, sizeof( FreeRTOS_Socket_t ) );
+
+    listLIST_ITEM_CONTAINER_ExpectAndReturn( &( xSocket.xBoundSocketListItem ), ( struct xLIST * ) ( uintptr_t ) 0x11223344 );
```

Comment:
Using ExpectAndReturn for all of these list operations seems very cumbersome. A test double for the list api would be a less brittle option.

Reply:
Sometimes we want lists to return values without initializing all the values in the list. The expectAndReturn serves that purpose well. I can Pass a list which is just a pointer without any memory but still get a value out of it.
Additionally, we have this across all the files in unit-tests. We'll need a separate PR to address this issue.